packages/protocol/src/abstracts/BaseVault.sol

```diff
@@ -970,4 +972,48 @@ abstract contract BaseVault is ERC20, SystemAccessControl, PausableVault, VaultP
       revert BaseVault__checkRebalanceFee_excessFee();
     }
   }
+
+  /// @inheritdoc IVault
+  function harvest(
+    Strategy strategy,
+    ILendingProvider provider,
+    bytes memory data
+  )
+    external
+    hasRole(msg.sender, HARVESTER_ROLE)
+  {
+    //check provider is valid
+    if (!_isValidProvider(address(provider))) {
+      revert BaseVault__harvest_invalidProvider();
+    }
+
+    //collect rewards from provider
+    _harvest(provider, data);
+
+    IERC20 token = IERC20(provider.getHarvestToken());
+
+    //strategy 1 = convert rewards to collateral
+    if (strategy == Strategy.ConvertToCollateral) {
+      //check if reward token is asset
+      if (address(token) != asset()) {
+        //swap for collateral
+      }
+    }
+
+    //strategy 2 = repay debt
+    /// @dev only implemented in BorrowingVault
+    if (strategy == Strategy.RepayDebt) {
+      revert BaseVault__harvest_strategyNotImplemented();
+    }
+
+    //strategy 3 = distribute rewards
+    if (strategy == Strategy.Distribute) {}
```

Comment:
Wouldn't it be more flexible if this logic is in the provider?

Reply:
Yes, it would. I actually was thinking about changing it because compoundV3 has a claimTo function which can be leveraged if we know the strategy. Well spotted !

Reply:
While implementing this, I realised it's not as useful as I thought to have this logic in the provider as it is in the vault. The claimTo function of CompoundV3 cannot be leveraged (we alway need to claimTo(address(vault)):
strategy 1: claim to vault to swap for collateral
strategy 2: claim to vault to swap for debt
strategy 3: cannot claim to all users to distribute rewards

In the general functionality of the harvest:
If we want to swap for collateral or debt (strategy 1 and 2) we will harvest and then swap. The strategy will still be in the ILendingProvider.harvest (and ILendingProvider.previewHarvest) in case we're able to choose which token to get rewards in. In those cases, the preview harvest will return the address[] tokens and uint256[] amounts according to the strategy (ex. previewHarvest will return the collateralAsset and amount if the strategy is convertToCollateral).

The logic present in the vault will be the swap (depending on the strategy and values returned by previewHarvest). If we put the swap logic inside the provider, we will be repeating a lot of code. 

Will do this changes in this PR.

Let me know your thoughts on this.